### PR TITLE
Changed comment on ScaledValue

### LINKED ISSUE #50564: Sign error in comment on resource.ScaledValue
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
I invoked `ScaledValue(-9)` on a resource.Quantity, and got a result that was multiplied by a factor of 10^9.

**What you expected to happen**:
I expected to get a value that was multiplied by a factor of 10^-9, because the comment says "ScaledValue returns the value of ceil(q * 10^scale)".

**How to reproduce it (as minimally and precisely as possible)**:
```
q, _ := resource.ParseQuantity("10")
i := q.ScaledValue(-1)
```
print `i`.  You will find it is 100, not 1.

**Anything else we need to know?**:
Look at the meaning of the `AsScaledInt64` method invoked in the integer case.  This method returns the quantity expressed as a multiple of 10^scale.  10 is 100 tenths.

**Environment**:
- Kubernetes version (use `kubectl version`): master

### PR DESCRIPTION
**Pull request by josiahbjorgaard**
**What type of PR is this?**
/kind documentation

**What this PR does / why we need it**: Changes incorrect comment ```ScaledValue returns the value of ceil(q * 10^scale);``` to ```ScaledValue returns the value of ceil(q / 10^scale);```. Adds an additional test of ScaledValue outside of named special cases.

**Which issue(s) this PR fixes**:

Fixes #50564

**Special notes for your reviewer**:

I tested the ScaledValue function with the following results to confirm the comment change:

With `q` as a `Quantity`, from `ScaledValueResult=q.ScaledValue(ScaledValueInput)` :
```
q.i.value=3,q.i.scale=-6,ScaleValueInput=-10,ScaledValueResult=30000
q.i.value=3,q.i.scale=-6,ScaleValueInput=-8,ScaledValueResult=300
q.i.value=3,q.i.scale=-6,ScaleValueInput=-6,ScaledValueResult=3
q.i.value=3,q.i.scale=-6,ScaleValueInput=-4,ScaledValueResult=1
q.i.value=3,q.i.scale=-6,ScaleValueInput=-2,ScaledValueResult=1
q.i.value=3,q.i.scale=-6,ScaleValueInput=0,ScaledValueResult=1
q.i.value=3,q.i.scale=0,ScaleValueInput=-4,ScaledValueResult=30000
q.i.value=3,q.i.scale=0,ScaleValueInput=-2,ScaledValueResult=300
q.i.value=3,q.i.scale=0,ScaleValueInput=0,ScaledValueResult=3
q.i.value=3,q.i.scale=0,ScaleValueInput=2,ScaledValueResult=1
q.i.value=3,q.i.scale=0,ScaleValueInput=4,ScaledValueResult=1
q.i.value=3,q.i.scale=0,ScaleValueInput=0,ScaledValueResult=3
```
**Does this PR introduce a user-facing change?**: NONE

```release-note

```

### CODE REVIEW

**Review by MikeSpreitzer — APPROVED** [review]

**Comment by cheftako** [comment]
Could you enhance https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go#L1010-L1032 with the examples you have where the expected result isn't 1?
Otherwise seems good.

**Comment by MikeSpreitzer** [comment]
@cheftako : huh?  There already are such cases, such as https://github.com/kubernetes/kubernetes/blob/b130ff450232eaf9100548aa08295168871a4ee9/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go#L1020

**Review by MikeSpreitzer — APPROVED** [review]

**Comment by MikeSpreitzer** [comment]
+1

**Comment by josiahbjorgaard** [comment]
@cheftako @MikeSpreitzer To add to this, the tests check for correct scaling when `q.i.value=1` which checks the `ceil` functionality of `ScaledValue` by rounding, e.g., 0.1 to 1. However, I noticed that the tests don't seem to cover `q.d.Dec != nil`, which is a large part of the `ScaledValue` function.

> @cheftako : huh? There already are such cases, such as
> 
> https://github.com/kubernetes/kubernetes/blob/b130ff450232eaf9100548aa08295168871a4ee9/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go#L1020

**Comment by cheftako** [comment]
> Walter Fender : huh? There already are such cases, such as
> 
> https://github.com/kubernetes/kubernetes/blob/b130ff450232eaf9100548aa08295168871a4ee9/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go#L1020

Your samples seem a slight improvement over the tests in that you use values other than 1 and scales other than 10^3. As such I think adding a few of those is helpful.

**Comment by MikeSpreitzer** [comment]
@cheftako:
The existing tests have "scales other than 10^3".
I assume the reason you are interested in "values other than 1" is that ceiling of a fraction is 1.  The existing tests have expected values other than 1.
I do not understand what you think is missing from the existing tests.

**Comment by MikeSpreitzer** [comment]
@josiahbjorgaard : I discussed the testing issues with @cheftako .  The part I was missing is that he is arguing for better coverage of the input space, not just better coverage of the code paths.  That is why he is asking for test cases where the expected values are not simply powers of 10.  The ask here is to generalize `TestScaledValue` so that it does not always call `NewScaledQuantity(1, ...)` but rather the first argument to NewScaledQuantity is part of the test case.

Also, the implementation has special cases for multiplying by the 0, 1, 2, 3, 6, and 9 power of 10, so we should have test cases that do not hit those special cases.

**Comment by josiahbjorgaard** [comment]
@cheftako @MikeSpreitzer I'd be happy to submit a PR with enhanced testing for the ScaledValue function. I'd like to enhance beyond what you mention above. I'm unable to find tests for the case `q.d.Dec != nil` as defined in the code snippet below. While tests for the `scaledValue` function exist in `scale_int_test.go` they don't test the combination of methods in the call to `scaledValue` performed in the `ScaledValue` function when `q.d.Dec != nil`. I think that this should also be a part of the `ScaledValue` tests.

```
func (q *Quantity) ScaledValue(scale Scale) int64 {
    if q.d.Dec == nil {
        i, _ := q.i.AsScaledInt64(scale)
        return i
    }
    dec := q.d.Dec
    return scaledValue(dec.UnscaledBig(), int(dec.Scale()), int(scale.infScale()))
}
```

**Comment by MikeSpreitzer** [comment]
@josiahbjorgaard : you're right, there is even more expansion that could be done to the tests.  @cheftako : how about we make that the mission of a separate PR?

**Comment by cheftako** [comment]
I agree that a full scale improvement to improve the tests could(should) be a separate issue/pr. That having been said the usage of more examples to show the doc/comment change is correct suggests we should also capture those examples as part of the doc change. I'd like to see at least a few added to the existing tests to help clarify usage.

**Comment by MikeSpreitzer** [comment]
We already have tests that show that the comment needs fixing, as I recently noted in the original issue (https://github.com/kubernetes/kubernetes/issues/50564#issuecomment-511014423).  So I suggest that for this PR we simply use the existing structure of `TestScaledValue` and add one test case `{2, -2, 100*100}` which (a) avoids the special cases and (b) demonstrates the need to fix the comment, leaving bigger changes to the testing for the other PR.

**Comment by josiahbjorgaard** [comment]
@cheftako I added the test to this PR. It needs new lgtm label.

**Comment by MikeSpreitzer** [comment]
@josiahbjorgaard : we generally squash multi-commit PRs before merging them.  Now would be a good time to squash this one.

**Comment by cheftako** [comment]
I would strongly recommend squashing this commits in this PR. ("git rebase -i HEAD~3" will start the process of squashing the last 3 commits. Follow directions from there.)

**Comment by josiahbjorgaard** [comment]
@cheftako They are now squashed

**Review by MikeSpreitzer — APPROVED** [review]

**Comment by MikeSpreitzer** [comment]
This is languishing because @cheftako is not an approver here.
/assign smarterclayton

**Comment by MikeSpreitzer** [comment]
@josiahbjorgaard : https://github.com/kubernetes/kubernetes/pull/79738#issuecomment-508182613 says that this PR needs an approval from someone listed as an approver in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/OWNERS , and @cheftako is not there, so I assigned someone who is.

**Comment by josiahbjorgaard** [comment]
/assign @liggett 
Are you able to look at this PR?

**Comment by MikeSpreitzer** [comment]
I don't know why @josiahbjorgaard got a complaint about "liggett" (note different spelling).  Let me try.

/assign @liggitt

File: staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go]: The new comment is accurate, thanks.

An example in the comment to help with intuition might be useful, e.g.:
```
For example, `NewQuantity(1, DecimalSI).ScaledValue(Milli)` returns 1000.
```

**Comment by MikeSpreitzer** [comment]
FYI about that flake: https://github.com/kubernetes/kubernetes/issues/80528
/retest

**Comment by liggitt** [comment]
Fix in https://github.com/kubernetes/kubernetes/pull/81050 once it gets through the queue

**Pull request merged**
